scvi_tools_gui/app.py

```diff
@@ -0,0 +1,452 @@
+        return [], px.violin(adata.obs, y="n_genes_by_counts"), px.violin(adata.obs, y="total_counts")
+
+@app.callback(
+    Output("setup-anndata-status",  "children"),
+    Input("setup-anndata-submit", "n_clicks"),
+    State("batch_key", "value"),
+    State("obsm_key", "value"),
+    State("uns_key", "value"),
+)
+def setup_anndata_callback(n_clicks, batch_key, obsm_key, uns_key):
+    if n_clicks >= 1:
+        adata = scvi.data.read_h5ad("data.h5ad")
+        anndata_config = {
+
+            'batch_key' : batch_key,
+            'protein_expression_obsm_key' : obsm_key,
+            'protein_names_uns_key' : uns_key,
+        }
+        anndata_config = {key : anndata_config[key] for key in anndata_config  if anndata_config[key] and anndata_config[key] != 'None'}
+        json.dump(anndata_config, open('anndata_config.json','w'))
+
+
+        return [dbc.Alert(
+            [
+                html.H4("Success!", className="alert-heading"),
+                html.Hr(),
+                html.P(
+                    str(adata),
+                    className="mb-0",
+                ),
+            ], dismissable=True, is_open=True,
+        )]
+
+
+
+@app.callback(
+    Output("train-model-status", "children"),
+    Input("train-button", "n_clicks"),
+    State("n_hidden", "value"),
+    State("lr", "value"),
+)
+def train_model_callback(n_clicks, n_hidden, lr):
+    if n_clicks >= 1:
+        adata = scvi.data.read_h5ad("data.h5ad")
+        anndata_config = json.loads(open("anndata_config.json", "r").read())
+        scvi.data.setup_anndata(
+            adata, 
+            **anndata_config
+        )
+        vae = scvi.model.SCVI(adata, n_hidden=n_hidden)
+        vae.train(lr=lr)
+        vae.save("my_model/")
+        return [dbc.Alert(
+            [
+                html.H4("Success!", className="alert-heading"),
+                html.Hr(),
+                html.P(
+                    "Trained model and saved it at 'my_model'."
+                ),
+
+            ], dismissable=True, is_open=True,
+        )]
+
+@app.callback(
+    Output("collapse", "is_open"),
+    [Input("collapse-button", "n_clicks")],
+    [State("collapse", "is_open")],
+)
+def toggle_collapse_callback(n, is_open):
+    if n:
+        return not is_open
+    return is_open
+
+@app.callback(Output("page-content", "children"), [Input("url", "pathname")])
+def render_page_content(pathname):
+    if pathname == "/":
+        return upload_page()
+    elif pathname == "/preprocess":
+        return preprocess_page()
+    elif pathname == "/setup-anndata":
+        return setup_anndata_page()
+    elif pathname == '/train-model':
+        return train_model_page()
+    return dbc.Jumbotron(
+        [
+            html.H1("404: Not found", className="text-danger"),
+            html.Hr(),
+            html.P(f"The pathname {pathname} was not recognised..."),
+        ]
+    )
+
+def run():
+    app.run_server()
+
+
+if __name__ == "__main__":
+    app.run_server()
```

Comment:
somehow we need to import app in the __init__ file, and write a CLI so that like `scvi_tools_gui launch` starts the app server.

you can see how to include a cli with poetry